src/CoinbaseWalletSDK.ts

```diff
@@ -92,6 +93,9 @@ export class CoinbaseWalletSDK {
     });
     this.setAppInfo(options.appName, options.appLogoUrl);
     this._relay.attachUI();
+
+    const serverUrl = window.encodeURIComponent(linkAPIUrl);
+    this._qrUrl = `${linkAPIUrl}/#/link?id=${this._relay.session.id}&secret=${this._relay.session.secret}&server=${serverUrl}&v=1`;
```

Comment:
let's create a dedicated class or function or module for the sake of DRY.
so i mean, refactoring `ScanQRCode` and `ScanQRBox` accordingly as well.

Reply:
also it'd be great to avoid manually building a query string, and instead to use an actual parser/generator - `URL`, or https://npmjs.com/qs stringify

Reply:
I caught up with the rainbow eng yesterday and he showed me how they're currently pulling the URI for WalletConnect and how they are fetching CBWSDK (via wagmi's connector). Given that they're using our web3 provider directly from wagmi, instead of importing our SDK library, I'm thinking we can follow the same pattern as WalletConnect and put this QR code uri as a public field on the `CoinbaseWalletProvider`. What do you guys think? @bangtoven @ljharb 

Adding screenshots because it's a private repo until release:

How they fetch WalletConnect URI:
<img width="1007" alt="Screen Shot 2022-04-08 at 8 40 31 AM" src="https://user-images.githubusercontent.com/15843286/162437671-da5f837b-f16d-4c16-a7e5-4ab32f7728f3.png">

What getProvider on wagmi does:
<img width="862" alt="Screen Shot 2022-04-08 at 8 41 26 AM" src="https://user-images.githubusercontent.com/15843286/162438526-0c890a29-84c8-40ec-a53b-3b00fe5ce136.png">

How they use CBWSDK:
<img width="1045" alt="Screen Shot 2022-04-08 at 8 44 31 AM" src="https://user-images.githubusercontent.com/15843286/162438044-cff3b1b2-aee3-44e9-a141-246100be1af6.png">

Reply:
that depends on how walletlink.org is implemented. Using the `/#/` convention is super archaic, and pushState should be used instead.

What I'd expect is that the `?` part is constructed with `qs` or `URLSearchParams`, and then appended to `https://www.walletlink.org/#/link`.

Reply:
ok. i've tried something like this:

<img width="539" alt="Screen Shot 2022-04-08 at 12 25 14 PM" src="https://user-images.githubusercontent.com/15843286/162483048-5a789687-a629-40d5-a1dc-5993cc4a78a6.png">

but the returning href looks like this: `https://www.walletlink.org/?id=...&secret=...&server=...&v=1#/link` where the `/#/link` is appended at the tail end. 

i'm seeing that these optional fragments are normally placed at the end of the URL, so now i'm thinking the original string is incorrect. thoughts on how to resolve this?
<img width="721" alt="Screen Shot 2022-04-08 at 12 26 47 PM" src="https://user-images.githubusercontent.com/15843286/162483308-3b6c110e-2743-40cd-858a-3e0ec1485c42.png">

Reply:
right - because it's behind a fragment, you can't use URLSearchParams on that link. You'd need to do some manual stuff like `new URL('.')`, and then you'd end up with `.?blah`, and then you'd need to slice out the `?blah`, and then append that manually to the actual URL string.

Reply:
sounds like the resulting URL construction might end up more difficult to read than the existing interpolated string - i might keep it the way it is for now if that's alright

Reply:
perhaps, but it'd be less bug-prone.

my primary suggestion is this:
```js
this._qrUrl = `${linkAPIUrl}/#/link${qs.stringify({
  id: this._relay.session.id,
  secret :this._relay.session.secret,
  server: serverUrl,
  v: 1,
})}`;
```
which i personally find far more readable.

Reply:
their current (hacky) method to get url shouldn't affect our design. 
seems wagmi is still using walletlink, so we have to make updates on them anyway. 
let's try to go with a better solution wrt separation of concerns. 
it appears to me that, walletlink.org connection related stuff doesn't belong to provider logic.

Reply:
I agree - but I'm also saying that *always*, building a URL query string should use something like qs or USP :-)

Reply:
@ljharb i tried using the qs library with the solution you suggested. however, when i stringify the following url (`https%3A%2F%2Fwww.walletlink.org`) it turns into this new url (`https%253A%252F%252Fwww.walletlink.org`) adding a few hex characters (namely, 25) that weren't present in the original string interpolation - might you know how to prevent this? it's breaking the QR linking when scanned with mobile

Reply:
Sorry if i wasn't clear. I don't mean the entire URL, i mean the query string part - just the part that starts with `?`. The rest would be manually prepended to the result.

Reply:
No worries, yeah i was just handling the query string part but there is a URL that goes into one of the search params (i.e. the `serverUrl` param which is this string `https%3A%2F%2Fwww.walletlink.org`) wondered if there was a way to work around it without manually writing just the last part out
```
this._qrUrl = `${linkAPIUrl}/#/link${qs.stringify({
  id: this._relay.session.id,
  secret :this._relay.session.secret,
  server: serverUrl,
  v: 1,
})}`;
```

Reply:
not that i know of, unfortunately :-/